cdap-app-fabric/src/main/java/co/cask/cdap/app/runtime/AbstractProgramRuntimeService.java

```diff
@@ -71,11 +90,91 @@ public RuntimeInfo run(Program program, ProgramOptions options) {
+    ProgramOptions optionsWithPlugins = createPluginSnapshot(optionsWithRunId, program.getId().getNamespace(), tempDir,
+                                                             program.getApplicationSpecification());
+    final RuntimeInfo runtimeInfo = createRuntimeInfo(runner.run(program, optionsWithPlugins), program);
+    programStarted(runtimeInfo, cleanUpTask);
     return runtimeInfo;
   }
 
+  private Runnable createCleanupTask(final File... resources) {
+    return new Runnable() {
+      @Override
+      public void run() {
+        for (File file : resources) {
+          if (file == null) {
+            continue;
+          }
+
+          try {
+            if (file.isDirectory()) {
+              DirUtils.deleteDirectoryContents(file);
+            } else {
+              file.delete();
+            }
+          } catch (Throwable t) {
+            LOG.warn("Exception when cleaning up resource {}", file, t);
+          }
+        }
+      }
+    };
+  }
+
+  /**
+   * Creates a local temporary directory for this program run.
+   */
+  private File createTempDirectory(Id.Program programId) {
+    File tempDir = new File(cConf.get(Constants.CFG_LOCAL_DATA_DIR),
+                            cConf.get(Constants.AppFabric.TEMP_DIR)).getAbsoluteFile();
+    File dir = new File(tempDir, String.format("%s.%s.%s.%s.%s",
+                                               programId.getType().name().toLowerCase(),
+                                               programId.getNamespaceId(), programId.getApplicationId(),
+                                               programId.getId(), UUID.randomUUID().toString()));
+    dir.mkdirs();
+    return dir;
+  }
+
+  /**
+   * Return the copy of the {@link ProgramOptions} including locations of plugin artifacts in it.
+   * @param options the {@link ProgramOptions} in which the locations of plugin artifacts needs to be included
+   * @param namespace Id of the Program's namespace
+   * @param tempDir Temporary Directory to create the plugin artifact snapshot
+   * @param appSpec program's Application Specification
+   * @return the copy of the program options with locations of plugin artifacts included in them
+   */
+  private ProgramOptions createPluginSnapshot(ProgramOptions options, Id.Namespace namespace, File tempDir,
+                                              @Nullable ApplicationSpecification appSpec) {
+    // appSpec is null in an unit test
+    if (appSpec == null) {
+      return options;
+    }
+
+    Id.Artifact appArtifactId = Id.Artifact.from(namespace, appSpec.getArtifactId());
+    ImmutableMap.Builder<String, String> builder = ImmutableMap.builder();
+    builder.putAll(options.getArguments().asMap());
+    for (Map.Entry<String, Plugin> pluginEntry : appSpec.getPlugins().entrySet()) {
+      Plugin plugin = pluginEntry.getValue();
+      try {
+        Map.Entry<ArtifactDescriptor, PluginClass> artifactEntry = artifactRepository.getPlugin(
```

Comment:
use `artifactRepository.getArtifact(pluginArtifactId)` instead of adding this method